src/effects/effectslot.cpp

```diff
@@ -268,3 +272,62 @@ void EffectSlot::slotEffectMetaParameter(double v, bool force) {
         pParameterSlot->onEffectMetaParameterChanged(v, force);
     }
 }
+
+QDomElement EffectSlot::toXml(QDomDocument* doc) const {
+    QDomElement effectElement = doc->createElement(EffectXml::Effect);
+    if (m_pEffect == nullptr) {
+        return effectElement;
+    }
+
+    QDomElement metaKnobElement = doc->createElement(EffectXml::EffectMetaParameter);
+    XmlParse::addElement(*doc, effectElement,
+                         EffectXml::EffectMetaParameter,
+                         QString::number(m_pControlMetaParameter->get()));
+    EffectManifest manifest = m_pEffect->getManifest();
+    XmlParse::addElement(*doc, effectElement,
+                         EffectXml::EffectId, manifest.id());
+    XmlParse::addElement(*doc, effectElement,
+                         EffectXml::EffectVersion, manifest.version());
+
+    QDomElement parametersElement = doc->createElement(EffectXml::ParametersRoot);
+
+    for (const auto& pParameter : m_parametersById) {
```

Comment:
> Since m_parametersById is only used in serializing/deserializing, I wonder if it makes sense to build it on the fly instead of storing at the class level (which requires keeping it in sync as new effects are loaded).

I don't think so. EffectSlot::loadEffectSlotFromXml needs to check that the parameter IDs in the XML match IDs in the effect's manifest. m_parametersById is built in EffectSlot::loadEffect when the Effect is loaded is loaded from the XML, which happens before EffectSlot::loadEffectSlotFromXml is called. (EffectSlot::loadEffect is called by EffectChainSlot::slotChainEffectChanged when the EffectChain emits its effectChanged signal in EffectChain::addEffect, which is called from EffectChain::createFromXml).

> Also, why a QMap? did you want the items to be sorted by ID in the XML (instead of being in the parameter order in the manifest)?

I used a QMap to pair the EffectParameterSlotBasePointers with the corresponding manifests' ID (which is a QString). EffectSlot::loadEffectSlotFromXml needs to lookup an EffectParameterSlotBasePointer from the ID string in the XML. It would be a good idea to serialize them in order when writing the XML though. Then when support for users rearranging parameters is implemented that could be represented in the XML file with the order of the `<Parameter>` elements.